### PR TITLE
GraalJs based Sequence handling with Configurable Engine Switch

### PR DESCRIPTION
**Pull request by shanggeeth**
### Proposed changes in this pull request

- Adaptive Authentication templates works with GraalJs
- Script engines can be switched before Identity Server Start Up

### Related Pull Requests

- https://github.com/wso2/product-is/pull/11447
- https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/66

### CODE REVIEW

**Comment by somindatommy** [comment]
@shanggeeth shall we remove the unwanted information in the PR description and update it properly? WDYT?

**Comment by shanggeeth** [comment]
> @shanggeeth shall we remove the unwanted information in the PR description and update it properly? WDYT?

Yes, I have removed the unnecessary information

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph;
+
+import org.wso2.carbon.identity.application.authentication.framework.internal.FrameworkServiceDataHolder;
+
+/**
+ * Singleton to get the appropriate Javascript Proxy Object Wrapper Factory.
+ */
+public class JsWrapperFactorySingleton {
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:26]: `JsWrapperFactoryProvider` ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private JsWrapperFactorySingleton() {
+        if (FrameworkServiceDataHolder.getInstance().
+                getJsGraphBuilderFactory() instanceof JsPolyglotGraphBuilderFactory) {
+            jsWrapperFactory = new GraalJsWrapperFactory();
+        } else {
+            jsWrapperFactory = new NashornJsWrapperFactory();
+        }
+    }
+
+    public static JsWrapperFactory getInstance() {
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:41]: `getFactory()` ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph;
+
+import org.wso2.carbon.identity.application.authentication.framework.internal.FrameworkServiceDataHolder;
+
+/**
+ * Singleton to get the appropriate Javascript Proxy Object Wrapper Factory.
+ */
+public class JsWrapperFactorySingleton {
+
+    private static JsWrapperFactorySingleton jsWrapperFactorySingleton = null;
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:28]:
```suggestion
    private static JsWrapperFactorySingleton jsWrapperFactorySingleton = new JsWrapperFactorySingleton();
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public static JsWrapperFactory getInstance() {
+        if (jsWrapperFactorySingleton == null) {
+            jsWrapperFactorySingleton = new JsWrapperFactorySingleton();
+        }
+        return jsWrapperFactory;
+    }
+
+    public static void clearInstance() {
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:48]: No need to clear the instance.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/NashornJsWrapperFactory.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.context.TransientObjectWrapper;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+
+import java.util.Map;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Factory to create a Javascript Object Wrappers for GraalJs execution.
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/NashornJsWrapperFactory.java:39]:
```suggestion
 * Factory to create a Javascript Object Wrappers for Nashorn execution.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.context.TransientObjectWrapper;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+
+import java.util.Map;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Interface specific Js Wrapper Factory.
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java:37]:
```suggestion
 * Interface to build Js wrapped objects.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsPolyglotGraphBuilderFactory.java [hunk 7]
```diff
@@ -18,76 +18,70 @@
...
-        Bindings engineBindings = engine.getBindings(ScriptContext.ENGINE_SCOPE);
+        Value engineBindings = context.getBindings(FrameworkConstants.JSAttributes.POLYGLOT_LANGUAGE);
         Map<String, Object> persistableMap = new HashMap<>();
-        engineBindings.forEach((key, value) -> persistableMap.put(key, FrameworkUtils.toJsSerializable(value)));
-        context.setProperty(JS_BINDING_CURRENT_CONTEXT, persistableMap);
-    }
-
-    public ScriptEngine createEngine(AuthenticationContext authenticationContext) {
+        engineBindings.getMemberKeys().forEach((key) -> {
+            Value keybinding = engineBindings.getMember(key);
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsPolyglotGraphBuilderFactory.java:67]:
```suggestion
            Value binding = engineBindings.getMember(key);
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java [hunk 8]
```diff
@@ -1,32 +1,32 @@
 /*
- *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java:2]: Is this file actually created in 2020?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java [hunk 9]
```diff
@@ -1,32 +1,32 @@
...
 
-import jdk.nashorn.api.scripting.AbstractJSObject;
 import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
 
 import java.io.Serializable;
+import java.util.Objects;
 
 /**
- * Represents the abstract class for all context objects
+ * Represents the abstract class for all context objects for GraalJs Execution.
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java:27]: Hope this is common for graal and nashorn both.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticationContext.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.base;
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticationContext.java:1]: License header

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseClaims.java [hunk 11]
```diff
@@ -1,31 +1,13 @@
...
- *  Unless required by applicable law or agreed to in writing,
- *  software distributed under the License is distributed on an
- *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- *  KIND, either express or implied.  See the License for the
- *  specific language governing permissions and limitations
- *  under the License.
- */
-
-package org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js;
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.base;
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseClaims.java:1]: License header

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseCookie.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.base;
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseCookie.java:1]: Check license header in all the new classes.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 13]
```diff
@@ -1359,8 +1364,8 @@ public static String getQueryStringWithFrameworkContextId(String originalQuerySt
         return idps;
     }
 
-    public static Map<String, AuthenticatorConfig> getAuthenticatedStepIdPs(StepConfig stepConfig,
-                                                                            Map<String, AuthenticatedIdPData> authenticatedIdPs) {
+    public static Map<String, AuthenticatorConfig>
+    getAuthenticatedStepIdPs(StepConfig stepConfig, Map<String, AuthenticatedIdPData> authenticatedIdPs) {
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:1368]: Reverted

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java (hunk 1 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:26]: Changed the name

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java (hunk 2 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:41]: Introduced `getFactory()` to return `JsWrapperFactory` and changed `getInstance()` to return `JsWrapperFactoryProvider`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java (hunk 4 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactorySingleton.java:48]: This is required while running the tests

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/NashornJsWrapperFactory.java (hunk 5 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/NashornJsWrapperFactory.java:39]: changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java (hunk 6 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java:37]: Modified to `to create`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java (hunk 8 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java:2]: This is the composite of both `AbstractJsContextMemberObject` classes
What year should be kept ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java (hunk 9 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/AbstractJSContextMemberObject.java:27]: Both, Corrected in other places as well

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticationContext.java (hunk 10 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticationContext.java:1]: Added, corrected in other places as well

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsAuthenticatedUser.java [hunk 14]
```diff
@@ -18,175 +18,11 @@
...
-                UserRealm userRealm = realmService.getTenantUserRealm(usersTenantId);
-                return userRealm.getUserStoreManager().getRoleListOfUser(usernameWithDomain);
-            } catch (UserStoreException e) {
-                LOG.error("Error when getting role list of user: " + getWrapped(), e);
-            }
-        }
-        return ArrayUtils.EMPTY_STRING_ARRAY;
-    }
+public interface JsAuthenticatedUser  {
+    AuthenticatedUser getWrapped();
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsAuthenticatedUser.java:27]: new line before  method and method doc

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticatedUser.java [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/base/JsBaseAuthenticatedUser.java:2]: when moving the files, we don't need to change the year

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 [hunk 16]
```diff
@@ -2458,6 +2458,9 @@
     </JWTValidatorConfigs>
 
     <AdaptiveAuth>
+        <!--Config for Script Engine Selection-->
```
Comment (senthalan) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:2461]:
```suggestion
        <!--Config for Script Engine Selection, Supported Values: nashorn, graaljs-->
```
let' add supported values as well

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/pom.xml [hunk 17]
```diff
@@ -67,6 +88,46 @@
                                     <overWrite>true</overWrite>
                                     <outputDirectory>${basedir}/src/main/resources/runtimes/cxf3</outputDirectory>
                                 </artifactItem>
+                                <artifactItem>
+                                    <groupId>org.graalvm.sdk</groupId>
+                                    <artifactId>graal-sdk</artifactId>
+                                    <version>${graalvm.version}</version>
+                                    <type>jar</type>
+                                    <overWrite>true</overWrite>
+                                    <outputDirectory>${basedir}/src/main/resources/runtimes/lib</outputDirectory>
```
Comment (senthalan) [inline features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/pom.xml:97]: let's change the outputDirectory
`${basedir}/src/main/resources/graal`

Change in other places a well.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/testng.xml [hunk 18]
```diff
@@ -18,13 +18,16 @@
...
+<!--        <parameter name="log-level" value="debug"/>-->
         <classes>
             <class name="org.wso2.carbon.identity.application.authentication.framework.ApplicationAuthenticationServiceTest"/>
             <class name="org.wso2.carbon.identity.application.authentication.framework.AbstractApplicationAuthenticatorTest"/>
 
             <class name="org.wso2.carbon.identity.application.authentication.framework.handler.hrd.impl.DefaultHomeRealmDiscovererTest"/>
             <class name="org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.DefaultAuthenticationRequestHandlerTest"/>
+            <class name="org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.DefaultRequestCoordinatorTest"/>
+            <class name="org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.JITProvisioningPostAuthenticationHandlerTest"/>
+            <class name="org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.PostAuthAssociationHandlerTest"/>
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/testng.xml:30]: Seems like duplicate entries ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 19]
```diff
@@ -2397,6 +2404,75 @@ public static Object toJsSerializable(Object value) {
         return value;
     }
 
+    public static Object toJsSerializableGraal (Object value) {
+        if (value instanceof Serializable) {
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:2408]: new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 20]
```diff
@@ -2418,8 +2494,41 @@ public static Object fromJsSerializable(Object value, ScriptEngine engine) throw
         return value;
     }
 
+    public static Object fromJsSerializableGraal (Object value, Context context) {
+        if (value instanceof GraalSerializableJsFunction) {
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:2498]: new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java [hunk 21]
```diff
@@ -775,4 +778,15 @@ private AuthenticatorConfig getAuthenticatorConfig(String name) {
         }
         return authConfig;
     }
+
+    private JsGraphBuilderFactory createJsGraphBuilderFactory() {
+        String scriptEngineName = IdentityUtil.getProperty("AdaptiveAuth.ScriptEngine");
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:783]: let's move to constant these to `AdaptiveAuth.ScriptEngine` and `graaljs`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsPolyglotGraphBuilder.java [hunk 22]
```diff
@@ -348,83 +240,47 @@ public Object evaluate(AuthenticationContext authenticationContext, Serializable
...
 
-    class ExecuteStepFunctionProxy implements ProxyExecutable {
+    /**
+     * Adds a function to show a prompt in Javascript code.
+     *
+     * @param templateId Identifier of the template
+     * @param parameters parameters
+     */
+    @SuppressWarnings("unchecked")
+    public void addShowPrompt(String templateId, Object... parameters) {
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsPolyglotGraphBuilder.java:261]: Another child class specific method is used in this method

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Interface to create Js Wrapper objects.
+ */
+public interface JsWrapperFactory {
+
+    JsAuthenticatedUser createJsAuthenticatedUser(AuthenticatedUser authenticatedUser);
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsWrapperFactory.java:41]: Added

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsAuthenticatedUser.java (hunk 14 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsAuthenticatedUser.java:27]: Changed and added method docs to similar interface methods

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 (hunk 16 above)
Comment (shanggeeth) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:2461]: Added

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/pom.xml (hunk 17 above)
Comment (shanggeeth) [reply features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/pom.xml:97]: Modified

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/testng.xml (hunk 18 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/testng.xml:30]: Removed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 19 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:2408]: Modified

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java (hunk 21 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:783]: Modified

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 20 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:2498]: Modified

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java (hunk 21 above)
Comment (shanggeeth) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:783]: Modified in all related scenarios

**Review by senthalan — APPROVED** [review]

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**